### PR TITLE
chore: SinkReplyBuilder2 with vec batching

### PR DESCRIPTION
**Pull request by dranikpg**
The idea:

```c++
ReplyScope scope(); // tells underlying reply builder it's safe to snatch long strings as iovecs

StartArray(n)
for i = 1, n {
  SendStr(arr[i])
}
```

and this way we can just write send code that internally turns into the same logic as in SendStrArr or SendMGetResponse (we also won't need custom methods)

### CODE REVIEW

**Comment by romange** [comment]
Nice. Why it is not enough to maintain the state with `StartArray` ?

**Comment by dranikpg** [comment]
> Nice. Why it is not enough to maintain the state with StartArray ?

1. Not all array sends guarantee pointer validity, we can check it in the future
2. I also want to batch non array operations that send protocol stuff

**Comment by dranikpg** [comment]
This is more or less what I used in my MVP

**Comment by romange** [comment]
it does not have enough comments for me to understand the meaning of helper functions, nor it has any code related to RESP protocol. How does it work?

File: src/facade/reply_builder.h [hunk 1]
```diff
@@ -168,6 +167,56 @@ class SinkReplyBuilder {
...
+  void Flush();        // Send all accumulated data and reset to clear state
+  void FinishScope();  // Called when scope ends
+
+  char* ReservePiece(size_t size);        // Reserve size bytes from buffer
+  void CommitPiece(size_t size);          // Mark size bytes from buffer as used
+  void WritePiece(std::string_view str);  // Reserve + memcpy + Commit
+
+  void WriteRef(std::string_view str);  // Add iovec bypassing buffer
+
+  bool IsInBuf(const void* ptr) const;
```
Comment (romange) [inline src/facade/reply_builder.h:206]: what  IsInBuf checks?

**Comment by dranikpg** [comment]
> it does not have any code related to RESP protocol. How does it work?

I'll update RedisReplyBuilder accordingly, this is a SinkReplyBuilder replacement 🤨

File: src/facade/reply_builder.h (hunk 1 above)
Comment (dranikpg) [reply src/facade/reply_builder.h:206]: if ptr is in buffer

**Review by romange — APPROVED** [review]

**Pull request merged**